### ISSUE TITLE
Failed to deserialize data

### ISSUE THREAD

**Issue by syed72** [shyiko/mysql-binlog-connector-java#51]
Stack Trace 
Exception: 
com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1433181420000, eventType=UPDATE_ROWS, serverId=478847, headerLength=19, dataLength=183, nextPosition=38149950, flags=0}
at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:140)
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:116)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:501)
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:328)
	at com.github.shyiko.mysql.binlog.BinaryLogClient$4.run(BinaryLogClient.java:425)
	at java.lang.Thread.run(Thread.java:744)
Caused by: java.net.SocketException: Socket closed
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:150)
	at java.net.SocketInputStream.read(SocketInputStream.java:121)
	at com.github.shyiko.mysql.binlog.io.BufferedSocketInputStream.read(BufferedSocketInputStream.java:51)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readWithinBlockBoundaries(ByteArrayInputStream.java:192)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:174)
	at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readLong(ByteArrayInputStream.java:57)
	at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeCell(AbstractRowsEventDataDeserializer.java:121)
	at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:96)
	at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserializeRows(UpdateRowsEventDataDeserializer.java:72)
	at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:58)
	at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:33)
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:134)
	... 5 more

Using 0.1.1 version of mysql-binlog-connector-java

public void onCommunicationFailure(BinaryLogClient client, Exception ex) {
	System.out.println("Communication failure ,Exception: "+ex.printStackTrace());
}

I'm getting above exception when any update/write/delete event occurs, saying not able to deserialize the event data. 
As you suggested, I tried setting BinaryLogClient.setServerId(uniqueid) and also keepAliveInterval to 60 sec, but still exception persists

**Comment by shyiko** [comment]
Hello Syed,

Please read through [#21](https://github.com/shyiko/mysql-binlog-connector-java/issues/21). If that does not help please specify version of MySQL server and OS.  

Thank you.

**Comment by syed72** [comment]
Hi shyiko,

I've gone through above issue, but it doesn't helped me, MySQL version is 5.5.21 and centos. 

Just now same exception occured for Write rows event
com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1433181448000, eventType=WRITE_ROWS, serverId=478847, headerLength=19, dataLength=124, nextPosition=46113472, flags=0}
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:140)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:116)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:501)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:328)
        at com.github.shyiko.mysql.binlog.BinaryLogClient$4.run(BinaryLogClient.java:425)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.io.EOFException
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:180)
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readInteger(ByteArrayInputStream.java:46)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeCell(AbstractRowsEventDataDeserializer.java:147)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:96)
        at com.github.shyiko.mysql.binlog.event.deserialization.WriteRowsEventDataDeserializer.deserializeRows(WriteRowsEventDataDeserializer.java:65)
        at com.github.shyiko.mysql.binlog.event.deserialization.WriteRowsEventDataDeserializer.deserialize(WriteRowsEventDataDeserializer.java:56)
        at com.github.shyiko.mysql.binlog.event.deserialization.WriteRowsEventDataDeserializer.deserialize(WriteRowsEventDataDeserializer.java:32)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:134)

**Comment by shyiko** [comment]
Could you please try latest (0.2.1) release to weed out some possibilities? Also, is it possible for you copy-and-paste part which configures & connects BinaryLogConnector so that I could try same thing on my end? Thank you.

**Comment by syed72** [comment]
Exception e: com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1434531671000, eventType=UPDATE_ROWS, serverId=19830019, headerLength=19, dataLength=813, nextPosition=1101009, flags=0}
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:140)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:116)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:501)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:328)
        at com.github.shyiko.mysql.binlog.BinaryLogClient$4.run(BinaryLogClient.java:425)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.io.EOFException
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:180)
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readLong(ByteArrayInputStream.java:57)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeCell(AbstractRowsEventDataDeserializer.java:117)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:96)
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserializeRows(UpdateRowsEventDataDeserializer.java:73)
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:58)
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:33)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:134)
        ... 5 more

I'm getting above kind of exceptions in some cases, 
I tried to read actual bin log at the position where i got this exception on the master,
after i read got this below data binlog position at which exception occured is  1100177

/*!40019 SET @@session.max_insert_delayed_threads=0*/;
/*!50003 SET @OLD_COMPLETION_TYPE=@@COMPLETION_TYPE,COMPLETION_TYPE=0*/;
DELIMITER /*!*/;
# at 1100177
#150617  2:00:02 server id 19830019  end_log_pos 0      Start: binlog v 4, server v 5.5.21-log created 150617  2:00:02
# at 1100177
#150617  2:01:11 server id 19830019  end_log_pos 1101009        Update_rows: table id 11575885 flags: STMT_END_F
### Row event for unknown table #11575885DELIMITER ;
# End of log file
ROLLBACK /* added by mysqlbinlog */;
/*!50003 SET COMPLETION_TYPE=@OLD_COMPLETION_TYPE*/;

**Comment by shyiko** [comment]
I would expect "Row event for unknown table ..." if you start reading from the position of the row event itself (and not from preceding TABLE_MAP event). I haven't tried it myself, though, so this is just a hunch.

**Comment by shyiko** [comment]
@syed72 any updates?

**Comment by syed72** [comment]
Now i moved to latest(0.2.1), still the above exception occurs, but its valid one as i mentioned in binary log with "Row event for unknown table " , and i skip that event and move to next event, from past 2 days, it was running fine, 

But today on one event, the binlog client has been struck, it not moving to next event, but other binary log clients are running fine even this deserialization failure occurs.

Exception:
Listener: Communication failure, BinLog: bin.007812(binlog name) 44598523(binlog pos) Exception e: com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1434776949000, eventType=UPDATE_ROWS, serverId=19830109, headerLength=19, dataLength=557, nextPosition=44599099, flags=0}
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:180)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:148)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:595)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:376)
        at com.github.shyiko.mysql.binlog.BinaryLogClient$4.run(BinaryLogClient.java:516)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.io.EOFException
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:182)
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.readInteger(ByteArrayInputStream.java:46)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeCell(AbstractRowsEventDataDeserializer.java:153)
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:99)
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserializeRows(UpdateRowsEventDataDeserializer.java:70)
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:58)
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:33)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:174)
        ... 5 more

Listener: Binary Log Client disconnect (onDisconnect)
Listener: Binary Log Client connect (onConnect)

After this exception, client trying to connect again, and trying to connect from the above binlog position and keeps on failing,

Note:
If i restart client, it starts from the next event of the above position. and running fine

This kind of above exceptions, i have to skip, But in some case, its not skipping this event,

Please suggest me for any workaround/solution

**Comment by syed72** [comment]
If i get below Exception, then client works fine by skipping the event which has problem,

Listener: Communication failure, BinLog: bin.001946 43916093 Exception e: com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1433181145000, eventType=UPDATE_ROWS, serverId=5234615, headerLength=19, dataLength=773, nextPosition=43916885, flags=0}
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:180)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:148)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:595)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:376)
        at com.github.shyiko.mysql.binlog.BinaryLogClient$4.run(BinaryLogClient.java:516)
        at java.lang.Thread.run(Thread.java:744)
Caused by: java.io.EOFException
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:182)
        at java.io.InputStream.read(InputStream.java:170)
        at java.io.InputStream.skip(InputStream.java:224)
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.skipToTheEndOfTheBlock(ByteArrayInputStream.java:208)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:176)
        ... 5 more

**Comment by shyiko** [comment]
Hi. Could you please show how do you configure (including event & lifecycle listeners) BinaryLogClient? Both exceptions you've provided do NOT cause BinaryLogClient to skip to the next event (you can see it [here](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L599)). You could force that kind of behavior by registering custom LifecycleListener with onCommunicationFailure doing something like `client.setBinlogPosition(((EventHeaderV4) ((EventDataDeserializationException) ex).getEventHeader()).getNextPosition())` but this is not something client does (nor should it be).

**Comment by shyiko** [comment]
Closing due to the lack of feedback.

**Comment by syed72** [comment]
Hi, Sorry for the delay

Below is the LifeCycleListener implementation

private class BinLogLifeCycleListener implements BinaryLogClient.LifecycleListener
	{
		public void onConnect(BinaryLogClient client) {
			LOG.info("Listener: Binary Log Client connect");
		}
		
		public void onCommunicationFailure(BinaryLogClient client, Exception ex) {
			LOG.info("Listener: Communication failure, BinLog: " +client.getBinlogFilename()+" "+client.getBinlogPosition()+" Exception e: "+StringUtils.stringifyException(ex));
			
			try
			{
				String exStr = ex.getMessage();
				LOG.info("exStr: "+exStr);//No I18N
				LOG.info("trace: "+ex.getCause());
				if (exStr.contains("Client requested master to start replication from impossible position")//No I18N
						|| exStr.contains("log event entry exceeded max_allowed_packet")
						|| exStr.contains("Failed to deserialize data of EventHeaderV4"))//No I18N
				{
					client.setBinlogPosition(Long.parseLong(binLogPos));
				}
				
				long servId = getServerId();

				if (serverId != servId)
				{
					LOG.info("server id changed due to takeover " + serverId + "old " + servId);//No I18N
					client.setBinlogFilename(null);
				}
				//client.setBinlogPosition(((EventHeaderV4) ((EventDataDeserializationException) ex).
					//		getEventHeader()).getNextPosition());
			}
			catch (Exception e)
			{
				LOG.info("Exception While getting serverid "+ StringUtils.stringifyException(e), e);
			}
			//LOG.info("going to connect after comm failure");
			//connect();	
		}
		
		public void onEventDeserializationFailure(BinaryLogClient client, Exception ex) {
			LOG.info("onEventDataDeserializationException Exception: " + StringUtils.stringifyException(ex));
			
		}
		
		public void onDisconnect(BinaryLogClient client) {
			LOG.info("Listener: Binary Log Client disconnect" );
		}
	}

please suggest me, if i am doing anything wrong.

**Comment by syed72** [comment]
How can i handle this unknown event type exceptions

onEventDataDeserializationException Exception: java.io.IOException: Unknown event type 114
        at com.github.shyiko.mysql.binlog.event.deserialization.EventHeaderV4Deserializer.getEventType(EventHeaderV4Deserializer.java:45)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventHeaderV4Deserializer.deserialize(EventHeaderV4Deserializer.java:35)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventHeaderV4Deserializer.deserialize(EventHeaderV4Deserializer.java:27)
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:143)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:595)
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:376)
        at com.github.shyiko.mysql.binlog.BinaryLogClient$4.run(BinaryLogClient.java:516)
        at java.lang.Thread.run(Thread.java:744)

**Comment by shyiko** [comment]
1 & 2. It seems like you won't be able to get EventHeaderV4 in case of EOF/Socket exception. EventDataDeserializationException
  gets automatically unwrapped in case of these two.
3. How can i handle this unknown event type exceptions? << You could supply custom [EventHeaderDeserializer](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/EventHeaderDeserializer.java) (default one is [EventHeaderV4Deserializer](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/EventHeaderV4Deserializer.java)) to EventDeserializer. But. This is really weird that you are getting event type 141 (to my knowledge, it's not part of stock version of MySQL). I would recommend you check your MySQL server installation. Do you have any custom extensions installed?

**Comment by syed72** [comment]
In 1 & 2 cases, how can i skip and move to next event?
as above said, i am configuring LifeCycleListener in that way, Whenever i get communication failure notification, how can i handle that, i mean, how to skip the EOF/socket exception during EventDataDeserializationException and move to next immediate event?

**Comment by shyiko** [comment]
The thing is - you shouldn't be handling these types of exceptions unless you need custom reconnection strategy (like exponential backoff) or whatnot. Getting persistent EOF/socket exceptions (by persistent I mean those which repeatedly happen on reconnect no matter how may times you retry) while reading event body is far from being OK. Something is seriously wrong with your MySQL installation. Even if you could skip to the next event in case of EOF/socket exceptions (like during event data processing) (which is absolutely wrong way to do it) what would you do if you get an exception while reading event header? There is no way to recover other then to reconnect from the last known position.

tl;dr you can't using stock mysql-binlog-connector-java. If need it - fork and change [BinaryLogClient's exception handling](https://github.com/shyiko/mysql-binlog-connector-java/blob/master/src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java#L596).

.3. As I wrote before: provide custom EventHeaderV4Deserializer (which sets eventType in case of unknown event to "UNKNOWN").
